### ISSUE TITLE
Code Refactor

### ISSUE THREAD

**Issue by jaitaiwan** [tcz/PHPTracker#44]
To reduce both the size of the repository as well as its complexity it's time to split some functionality out.

So far from my analysis of the code we need to split it out to at the very least 3 parts:
 - Bencoder
 - Process Forking
 - Tracking Code

Ideally I'd like to see it split out to:
 - Bencode
 - Process Forking
 - Tracker
   - Announce
   - Torrent
   - Client

But that might be a bit ambitious. I think the [feross/webtorrent](https://github.com/feross/webtorrent) repo is a good example of a separation of concerns and I'd like to get to a similar level of componentization in this one as well. 

To encourage contributing to other projects, it might also be good to look at using other more developed projects to fill the place of the existing Bencoder and Process Forking parts. At the moment I know that the Bencoder can likely be done in a lot less code as per my experiment: [bitbucket::centricwebestate/bencode](http://bitbucket.org/centricwebestate/bencode).

My main rational for moving the Process Forking code out is because in the case of systems like docker or even in use with systemd or upstart, sometimes we want process forking to be handled by an external system.

It also helps us to reduce reliance on a particular non-standard php library where some people might use an alternative to pcntl.

@genaside For me this means putting a hold on the other feature requests until this is complete. What do you think?

**Comment by genaside** [comment]
Why not just create a library for Phptracker to use? or is that what your actually saying? I think having a library will be doing the seeding and leeching while PHPtracker does the tracker stuff only. This will reduce the code of phptracker significantly. 

> To encourage contributing to other projects, it might also be good to look at using other more 
developed projects to fill the place of the existing Bencoder and Process Forking parts. 

Maybe porting.

> It also helps us to reduce reliance on a particular non-standard php library where some people might use an alternative to pcntl.

makes sense, I was really thinking of using pecl posix for the leecher but since it wasn't part of php main package i gave up on it. The leecher iam creating has two different functions for downloading one that uses forking and one that does not. Maybe when we split things out the code can support both pecl and pcntl?

>genaside For me this means putting a hold on the other feature requests until this is complete. What do you think?

The more i think about having library created the more i think its this best approach. Once the library is done the creation of bittorent based streaming media, file strorage, etc will be easy peasy.

Anyways, I can put a hold on what Iam doing, since the multiple file support is not ready. Yeah ok, just tell me of the things i can do.

**Comment by jaitaiwan** [comment]
I can pull out bencoding if you want to pull out forking?
